### PR TITLE
[FSDP2] Showed 2D MLP with colwise + colwise sharding

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126073
* #126070
* #126067

What is the point of this?
- FSDP2 currently only supports rowwise sharding (i.e. dim-0 sharding). This means that when combined with tensor parallelism, some linear weights are rowwise sharded twice, meaning that one matrix dim is sharded both by TP and FSDP while the other matrix dim is not sharded.
- For possible performance reasons (e.g. quantization block sizes), we may prefer to change those linear weights to have FSDP instead do colwise sharding (i.e. dim-1 sharding). However, implementing that in FSDP2 introduces some complexity. It may be simpler to transpose the weight and change TP to shard colwise. (More verbosely, instead of using TP rowwise sharding and FSDP colwise sharding, we transpose the weight, use TP colwise sharding, and FSDP rowwise sharding.)

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -932,6 +938,75 @@ def _test_train_parity_2d_mlp(
...
+            ref_model,
+            tp_mesh,
+            {"w1.weight": ColwiseParallel(), "w2.weight": RowwiseParallel()},
+        )
+        fully_shard(ref_model, mesh=dp_mesh)
+        ref_optim = torch.optim.AdamW(ref_model.parameters(), lr=1e-2)
+        model = parallelize_module(
+            model,
+            tp_mesh,
+            {"w1.weight": ColwiseParallel(), "w2.weight": ColwiseParallel()},
```
[the remark this answers is not in the document]
Comment (yifuwang) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:994]: > meaning that one matrix dim is sharded both by TP and FSDP while the other matrix dim is not sharded.

Would it make sense to just support this? Curious if this feels natural to folks (for me it kinda is). If so, how difficult would this be (if possible)?

[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:994]: @yifuwang That is the current way 2D works for FSDP2 (so it is supported, just missing the strided sharding placement that is being worked on, which only affects resharding).
